File: Find Minimum in Rotated Sorted Array.cpp
// Time: O(log n)
// Space: O(1)

class Solution {
public:
    int findMin(vector<int>& nums) {
        int n = nums.size();

        //if first element <= last element => either array is sorted and not rotated or there is just one element, return the first element
        if(nums[0] <= nums[n-1]) {
            return nums[0];
        }

        int l = 0, r = n-1;
        while(l <= r) {
            int mid = l + (r-l)/2;

            //if at any point there is just one element or the leftmost < rightmost, return leftmost
            if(nums[l] <= nums[r]) {
                return nums[l];
            }

            //if we reach last element OR when mid < both neighbors, return mid element
            else if(mid == n-1 || (mid != 0 && nums[mid] < nums[mid+1] && nums[mid] < nums[mid-1])) {
                return nums[mid];
            }

            //left sorted arr, min element has to be on the right
            else if(nums[mid] >= nums[l]) {
                l = mid+1;
            }

            //right sorted arr, min has to be on the left
            else {
                r = mid-1;
            }
        }
        return 0;
    }
};

File: find-minimum-in-rotated-sorted-array.java
// Time Complexity :O(logn)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach in three sentences only:
//  I used a binary search approach to find the min element in an array. 
// I start by calculating the middle index mid.
//  If nums[mid] is less than its nums[m-1] and mid!=0, it is a min, and I return nums[mid].otherwise i just check which side part of array is sorted by comparing nums[low]<nums[mid] it means left size is sorted it indicated our min defintle will be in rightside 
//  So, I update low = mid + 1, effectively ignoring the left half. Otherwise, I search in the left half
//  by setting high = mid - 1. This process continues, reducing the search space until the peak element is found.



class Solution {
    public int findMin(int[] nums) {
        int low=0;
        int high=nums.length-1;
        while(low <= high){
            if(nums[low] <= nums[high]){
                return nums[low];
            }else{
                int mid=low+(high-low)/2;
                //  if(low == high){
                //     return nums[low];
                // }
                   if(mid != 0 && nums[mid] < nums[mid-1]){
                    return nums[mid];
                }else{
                    //  if(mid == low){
                    //     low=mid+1;
                    //  }
                    if(nums[low] <= nums[mid]){
                       low=mid+1;
                    }else{
                        high=mid-1;
                    }
                }       
                  }
        }
        return 4567;
    }
}